### PR TITLE
Add the option of two factor auth to the admin

### PR DESCRIPTION
**Pull request by nburka**

### CODE REVIEW

File: Admin/components/AdminSite/Profile.php [hunk 1]
```diff
@@ -164,6 +167,40 @@ protected function buildInternal()
...
+				$this->ui->getWidget('google_2fa_enabled_note')->visible = true;
+			} else {
+				$qr_code_url = $google2fa->getQRCodeUrl(
+					$this->app->config->admin->google_2fa_domain,
+					$this->data_object->email,
+					$this->data_object->google_2fa_secret
+				);
+
+				$img_tag = new SwatHtmlTag('img');
+				$img_tag->src = 'https://chart.googleapis.com/chart'.
```
Comment (nburka) [inline Admin/components/AdminSite/Profile.php:197]: this is unfortunately deprecated. I think we'll need to replace with Bacon https://github.com/antonioribeiro/google2fa#baconqrcode

**Comment by nburka** [comment]
https://app.shortcut.com/silverorangeinc/story/37216/improve-login-for-emrap-admin

File: Admin/components/AdminSite/Google2fa.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	{
+		return new AdminLoginLayout($this->app, AdminLoginTemplate::class);
+	}
+
+	// }}}
+	// {{{ protected function initInternal()
+
+	protected function initInternal()
+	{
+		parent::initInternal();
```
Comment (nburka) [inline Admin/components/AdminSite/Google2fa.php:26]: add a check to make sure 2FA is enabled…

File: Admin/components/AdminSite/Profile.php [hunk 3]
```diff
@@ -164,6 +171,52 @@ protected function buildInternal()
...
+					)
+				);
+
+				$img_tag = new SwatHtmlTag('img');
+				$img_tag->alt = Admin::_('Two Factor Authentication QR Code');
+				$img_tag->src = 'data:image/png;base64, '.base64_encode(
+					$writer->writeString($qr_code_url)
+				);
+
+				// $img_tag->src = 'https://chart.googleapis.com/chart'.
```
Comment (nburka) [inline Admin/components/AdminSite/Profile.php:213]: remove this

File: Admin/components/AdminSite/Profile.php [hunk 4]
```diff
@@ -164,6 +220,60 @@ protected function buildInternal()
...
+				$this->data_object->save();
+
+				$qr_code_url = $two_fa->getQRCodeImageAsDataUri(
+					sprintf(
+						'%s (%s)',
+						$this->app->config->site->title,
+						$this->data_object->email
+					),
+					$this->data_object->two_fa_secret,
+					400
```
Comment (gauthierm) [inline Admin/components/AdminSite/Profile.php:252]: Extract 400 to a class constant. I assume this is image width.

File: Admin/components/AdminSite/Profile.php [hunk 5]
```diff
@@ -164,6 +220,60 @@ protected function buildInternal()
...
+						new ImagickImageBackEnd()
+					)
+				);
+
+				$img_tag = new SwatHtmlTag('img');
+				$img_tag->alt = Admin::_('Two Factor Authentication QR Code');
+				$img_tag->src = $qr_code_url;
+
+				$p_tag = new SwatHtmlTag('p');
+				$p_tag->class = 'two-factor-secret';
```
Comment (gauthierm) [inline Admin/components/AdminSite/Profile.php:267]: prefix css class with `admin-`

File: Admin/components/AdminSite/TwoFactorAuthentication.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	// }}}
+	// {{{ protected function validate2Fa()
+
+	protected function validate2Fa()
+	{
+		// strip all non numeric characters like spaces and dashes that people
+		// might enter (e.g. Authy adds spaces for readability)
+		$token = preg_replace(
```
Comment (gauthierm) [inline Admin/components/AdminSite/TwoFactorAuthentication.php:74]: this logic is copied in the profile page. It makes sense to move it to a separate class used in both places that wraps around `TwoFactorAuth`

File: Admin/components/AdminSite/profile.xml [hunk 7]
```diff
@@ -40,6 +47,20 @@
 					<widget class="SwatConfirmPasswordEntry" id="confirm_password" />
 				</widget>
 			</widget>
+			<widget class="SwatDisclosure" id="google_2fa">
+				<property name="title" translatable="yes">Two Factor Authentication</property>
```
[the remark this answers is not in the document]
Comment (gauthierm) [reply Admin/components/AdminSite/profile.xml:51]: This looks done below 👍

File: Admin/components/AdminSite/Profile.php [hunk 8]
```diff
@@ -103,6 +109,51 @@ protected function validate()
...
+			'',
+			$this->ui->getWidget('two_fa_token')->value
+		);
+
+		// The timestamp is used to make sure this, or tokens before this,
+		// can't be used to authenticate again. There's a "window" of token
+		// use and without this, someone could capture the code, and re-use it.
+		$two_fa = new TwoFactorAuth();
+		$success = $two_fa->verifyCode(
+			$this->app->session->user->two_fa_secret,
```
Comment (gauthierm) [inline Admin/components/AdminSite/Profile.php:136]: The profile page uses `$this->data_object` to refer to the account in most places. Does it make sense here too?

**Review by gauthierm — CHANGES_REQUESTED** [review]

**Review by gauthierm — APPROVED** [review]

**Pull request merged**